lib/Transforms/Scalar/ScalarReplAggregatesHLSL.cpp

```diff
@@ -3268,15 +3327,33 @@ static void ReplaceMemcpy(Value *V, Value *Src, MemCpyInst *MC) {
     if (TyV == TySrc) {
       if (V != Src)
         V->replaceAllUsesWith(Src);
+    } else if (!IsUnboundedArrayMemcpy(TyV, TySrc)) {
+      Value* DestVal = MC->getRawDest();
+      Value* SrcVal = MC->getRawSource();
+      if (!isa<BitCastInst>(SrcVal) || !isa<BitCastInst>(DestVal)) {
+        DXASSERT(0, "Encountered unexpected instruction sequence");
+        return;
+      }
+
+      BitCastInst *DestBCI = cast<BitCastInst>(DestVal);
+      BitCastInst *SrcBCI = cast<BitCastInst>(SrcVal);
+      if (!ArePointersToStructsOfIdenticalLayouts(DestBCI->getSrcTy(), SrcBCI->getSrcTy())) {
+        DXASSERT(0, "Can't handle structs of different layouts");
+        return;
+      }
+      if (DestBCI->hasOneUse() && SrcBCI->hasOneUse()) {
+        IRBuilder<> Builder(MC);
+        StructType *srcStTy = cast<StructType>(SrcBCI->getOperand(0)->getType()->getPointerElementType());
+        std::vector<unsigned> idxlist = { 0 };
+        CopyElementsOfStructsWithIdenticalLayout(Builder, DestBCI->getOperand(0), SrcBCI->getOperand(0), srcStTy, idxlist);
```

Comment:
Before doing the copy, could you also check if size of bytes copied by memcpy matches the struct size of source and destination structs.  LGTM otherwise.

Reply:
No need. If the types have different sizes, the compilation fails in earlier stages. I'll add a test to confirm that.